### PR TITLE
PyTorch Backend (LibTorch) Support

### PR DESCRIPTION
**Pull request by CoderHam**
**PyTorch (LibTorch) Backend:**
- [x] Process Input Tensors  (Support for multiple inputs included)
- [x] Execute Model (GPU + CPU)
- [x] Process Output Tensors (Support for multiple outputs included)

PS: The user has to ensure that the **model config file** and **request headers** contains the input and output names with the following naming convention: **\<name\>__\<index\>** (Start at 0).

### CODE REVIEW

File: src/backends/libtorch/torch_backend.cc [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "src/backends/libtorch/torch_backend.h"
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend.cc]: Backend should be pytorch, and "torch" is an old framework so I would use "libtorch". So src/backends/pytorch/libtorch_backend.so

File: src/backends/libtorch/torch_backend.cc [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+
+}  // namespace
+
+LibTorchBackend::Context::Context(
+    const std::string& name, const int gpu_device, const int max_batch_size)
+    : name_(name), gpu_device_(gpu_device), max_batch_size_(max_batch_size)
+{
+}
+
+LibTorchBackend::Context::Context(Context&& o)
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend.cc]: This move constructor is a legacy thing that we want to get rid of in new backends. Instead make contexts_ a unique_ptr vector. See how plan_backend handles it.

File: src/backends/libtorch/torch_backend.cc [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+
+  return Status::Success;
+}
+
+Status
+LibTorchBackend::Context::ValidateInputs(
+    const ::google::protobuf::RepeatedPtrField<ModelInput>& ios)
+{
+  for (const auto& io : ios) {
+    // For now, skipping the check if potential names is empty
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend.cc]: This comment doesn't make sense to me. It seems you are validating that the datatype is supported.

File: src/backends/libtorch/torch_backend.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    // batching is not supported.
+    int max_batch_size_;
+
+    // LibTorch model.
+    std::shared_ptr<torch::jit::script::Module> torch_model_;
+    std::vector<torch::jit::IValue> inputs_;
+    torch::Tensor outputs_;
+  };
+
+  std::vector<Context> contexts_;
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend.h]: See plan_bundle for how to handle the contexts_ variable.

File: src/servers/main.cc [hunk 5]
```diff
@@ -245,25 +245,25 @@ CheckPortCollision()
 {
   // Check if HTTP and GRPC have shared ports
   if ((std::find(http_ports_.begin(), http_ports_.end(), grpc_port_) !=
-       http_ports_.end()) && (grpc_port_ != -1) &&
-       allow_http_ && allow_grpc_) {
+       http_ports_.end()) &&
```
Comment (deadeyegoodwin) [inline src/servers/main.cc]: These changes seem unrelated to the PR and should go in a different PR

File: libtorch_test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (deadeyegoodwin) [inline libtorch_test/CMakeLists.txt]: The libtorch_test files should not be included in the PR since we don't want them on master.

File: src/backends/libtorch/torch_backend_c2.cc [hunk 7]
```diff
@@ -0,0 +1,322 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend_c2.cc]: The "c2" in the filename refers to "caffe2" so it is not a good name here. Perhaps libtorch_backend_pt.cc.

The point of this file is that it will actually be compiled into caffe2 when we rebuild it and so it is the only file that has to know anything about caffe2 includes, datatypes, etc. This was needed because caffe2 didn't have an API that TRTIS could use directly. If the libtorch API can be used directly (specifically, it does not require any protobufs to be passed as parameters, and does not require any datatypes as parameters that aren't available in the libtorch header) then you don't need this isolation (it all depends on the libtorch API, so we can discuss if you are unsure).

File: src/backends/libtorch/torch_backend_c2.h [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <set>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <ATen/ATen.h>
```
Comment (deadeyegoodwin) [inline src/backends/libtorch/torch_backend_c2.h]: Needing to include ATen implies that we do need to isolate libtorch (If ATen really needed to use the API?). So these includes should not appear in this header, only in torch_backend_c2.cc, since that will be compiled with pytorch.

File: src/core/constants.h [hunk 9]
```diff
@@ -42,6 +42,7 @@ constexpr char kTensorFlowGraphDefPlatform[] = "tensorflow_graphdef";
 constexpr char kTensorFlowSavedModelPlatform[] = "tensorflow_savedmodel";
 constexpr char kTensorRTPlanPlatform[] = "tensorrt_plan";
 constexpr char kCaffe2NetDefPlatform[] = "caffe2_netdef";
+constexpr char kLibTorchPtPlatform[] = "libtorch_pt";
```
Comment (deadeyegoodwin) [inline src/core/constants.h]: kPyTorchLibTorchPlatform = "pytorch_libtorch"

File: src/core/constants.h [hunk 10]
```diff
@@ -51,6 +52,7 @@ constexpr char kTensorRTPlanFilename[] = "model.plan";
 constexpr char kTensorFlowGraphDefFilename[] = "model.graphdef";
 constexpr char kTensorFlowSavedModelFilename[] = "model.savedmodel";
 constexpr char kCaffe2NetDefFilename[] = "model.netdef";
+constexpr char kLibTorchPtFilename[] = "model.pt";
```
Comment (deadeyegoodwin) [inline src/core/constants.h]: kPyTorchLibTorchFilename

File: src/core/model_config.h [hunk 11]
```diff
@@ -60,7 +60,8 @@ enum Platform {
   PLATFORM_CAFFE2_NETDEF = 4,
   PLATFORM_CUSTOM = 5,
   PLATFORM_ENSEMBLE = 6,
-  PLATFORM_ONNXRUNTIME_ONNX = 7
+  PLATFORM_ONNXRUNTIME_ONNX = 7,
+  PLATFORM_LIBTORCH_PT = 8
```
Comment (deadeyegoodwin) [inline src/core/model_config.h]: PLATFORM_PYTORCH_LIBTORCH

File: src/backends/libtorch/torch_backend.cc (hunk 1 above)
Comment (CoderHam) [reply src/backends/libtorch/torch_backend.cc]: Sure. Will change this

File: src/servers/main.cc (hunk 5 above)
Comment (CoderHam) [reply src/servers/main.cc]: Automatic changes due to formatting + rebase. I can remove these as these changes are already in master

File: src/backends/libtorch/torch_backend.cc (hunk 3 above)
Comment (CoderHam) [reply src/backends/libtorch/torch_backend.cc]: Removed

File: libtorch_test/CMakeLists.txt (hunk 6 above)
Comment (CoderHam) [reply libtorch_test/CMakeLists.txt]: Removed

File: src/backends/libtorch/torch_backend_c2.cc (hunk 7 above)
Comment (CoderHam) [reply src/backends/libtorch/torch_backend_c2.cc]: Workspace was not required for LibTorch. Refactored to fix this.

File: src/backends/pytorch/torch_backend.cc [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+    device_ = torch::Device(torch::kCPU);
+  } else {
+    device_ = torch::Device(torch::kCUDA, device_option.device_id)
+  }
+  try {
+    // lp_itr->second is the torch model path
+    torch_model_ = torch::jit::load(lp_itr->second, device_);
+    model_name_= Config().name();
+    max_batch_size_ = Config().max_batch_size();
+    gpu_device_ = gpu_device;
```
Comment (GuanLuo) [inline src/backends/pytorch/torch_backend.cc]: Aren't these variables members under Context struct? Should prefix them with `context->...`

File: src/backends/pytorch/torch_backend.cc [hunk 13]
```diff
@@ -0,0 +1,687 @@
...
+    case TYPE_STRING:
+    default:
+        return std::make_pair(false, type);
+  }
+
+  return std::make_pair(true, type);
+}
+
+const std::string
+DataTypeName(const DataType& dtype)
```
Comment (GuanLuo) [inline src/backends/pytorch/torch_backend.cc]: Don't need to define this function as you can use `DataType_Name()` directly (and you are already using it below)

File: src/backends/pytorch/torch_backend.cc [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+    std::vector<int64_t>* content_shape)
+{
+  // Initialize char* content[output_flat.nbytes()];
+  torch::DeviceType output_device = torch::kCPU;
+  try{
+    outputs_ = outputs_.to(output_device)
+    torch::Tensor output_flat = outputs_.flatten();
+    std::vector<float> outputs_vector;
+    for(int i=0;i<output_flat.sizes()[0];i++){
+      outputs_vector.push_back(output_flat[i].item().to<float>());
```
Comment (GuanLuo) [inline src/backends/pytorch/torch_backend.cc]: Seems to be an extra copy, now you are doing "`output_flat` -> `outputs_vector` -> `content`". You should be able to copy from `output_flat` to `content` directly

File: src/backends/pytorch/libtorch_backend.cc [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+    outputs_[op_index] = outputs_[op_index].to(output_device);
+    torch::Tensor output_flat = outputs_[op_index].flatten();
+    *byte_size = output_flat.nbytes();
+    *content = new char[*byte_size];
+
+    // Copy output into buffer
+    switch (dtype) {
+      case TYPE_UINT8: {
+        std::vector<uint8_t> outputs_vector;
+        for (int i = 0; i < output_flat.sizes()[0]; i++) {
```
Comment (GuanLuo) [inline src/backends/pytorch/libtorch_backend.cc]: You should be able to use template to make the following looks nicer.

File: src/backends/pytorch/libtorch_backend.cc [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+          outputs_vector.push_back(output_flat[i].item().to<double>());
+        }
+        memcpy(*content, &outputs_vector[0], output_flat.nbytes());
+      } break;
+    }
+
+    std::vector<float> outputs_vector;
+    for (int i = 0; i < output_flat.sizes()[0]; i++) {
+      outputs_vector.push_back(output_flat[i].item().to<float>());
+    }
```
Comment (GuanLuo) [inline src/backends/pytorch/libtorch_backend.cc]: Remove this, not being used anywhere.

File: src/backends/pytorch/libtorch_backend.cc [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+    // If 'payload' requested this output then copy it from
+    // 'content'. If it did not request this output then just
+    // skip it in the 'content'.
+    if ((payload.response_provider_ != nullptr) &&
+        payload.response_provider_->RequiresOutput(name)) {
+      void* buffer;
+      Status status = payload.response_provider_->AllocateOutputBuffer(
+          name, &buffer, expected_byte_size, content_shape);
+      if (status.IsOk()) {
+        memcpy(buffer, content + content_offset, expected_byte_size);
```
Comment (GuanLuo) [inline src/backends/pytorch/libtorch_backend.cc]: Extra memcpy. You copy once from output tensor to `content`, and then from `content` to `buffer`, should be able to copy directly from output tensor to `buffer`

File: src/backends/pytorch/libtorch_backend.cc (hunk 15 above)
Comment (GuanLuo) [reply src/backends/pytorch/libtorch_backend.cc]: Take a look at Tensor class's `data()` or `data_ptr()`

File: src/backends/pytorch/torch_backend.cc (hunk 14 above)
Comment (CoderHam) [reply src/backends/pytorch/torch_backend.cc]: Resolved to copy directly from `output_flat` -> `content`.

File: src/backends/pytorch/libtorch_backend.cc (hunk 15 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Yup. Figured a way to copy directly into memory using data_ptr()

Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Because of this, I did not need a template.

File: src/backends/pytorch/libtorch_backend.cc (hunk 16 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 18]
```diff
@@ -0,0 +1,745 @@
...
+
+    content_offset += expected_byte_size;
+  }
+
+  return Status::Success;
+}
+
+Status
+LibTorchBackend::Context::GetOutputTensor(
+    const std::string& name, const int& op_index, const DataType dtype, char** content,
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 19]
```diff
@@ -0,0 +1,745 @@
...
+    context->device_ = torch::Device(torch::kCPU);
+  } else {
+    context->device_ = torch::Device(torch::kCUDA, gpu_device);
+  }
+  try {
+    // lp_itr->second is the torch model path
+    context->torch_model_ = torch::jit::load(lp_itr->second, context->device_);
+    context->name_ = Config().name();
+    context->max_batch_size_ = Config().max_batch_size();
+    context->gpu_device_ = gpu_device;
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Yes. name, max_batch_size and gpu_device removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 20]
```diff
@@ -0,0 +1,745 @@
...
+{
+  cudaError_t cuerr;
+
+  // For a GPU context, determine the model file to use for device
+  // compute capability. CPU always uses the default model file.
+  std::string cc;
+  std::string cc_model_filename;
+  if (gpu_device == Context::NO_GPU_DEVICE) {
+    cc_model_filename = Config().default_model_filename();
+    LOG_INFO << "Creating instance " << instance_name << " on CPU using "
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 21]
```diff
@@ -0,0 +1,745 @@
...
+      type = torch::kFloat;
+      break;
+    case TYPE_FP64:
+      type = torch::kDouble;
+      break;
+    case TYPE_UINT16:
+    case TYPE_UINT32:
+    case TYPE_UINT64:
+    case TYPE_STRING:
+      break;
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed

File: src/backends/pytorch/libtorch_backend.cc [hunk 22]
```diff
@@ -0,0 +1,745 @@
...
+  return Status::Success;
+}
+
+Status
+LibTorchBackend::Context::GetOutputTensor(
+    const std::string& name, const int& op_index, const DataType dtype, char** content,
+    size_t* byte_size, std::vector<int64_t>* content_shape)
+{
+  torch::DeviceType output_device = torch::kCPU;
+  // TODO: Fix for supporting multiple outputs
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Fixed. Removed

File: src/backends/pytorch/libtorch_backend.h [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+
+    // The GPU index active when this context was created.
+    int gpu_device_;
+
+    // Maximum batch size to allow. NO_BATCHING indicates that
+    // batching is not supported.
+    int max_batch_size_;
+
+    // LibTorch model and variables that will be reset and used for every run
+    std::shared_ptr<torch::jit::script::Module> torch_model_;
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.h]: Old comment. It does not reset the model after each run anymore. Removed.

File: src/backends/pytorch/libtorch_backend.cc [hunk 24]
```diff
@@ -0,0 +1,745 @@
...
+  try {
+    auto model_outputs_ =
+        torch_model_->forward(inputs_).toTuple();
+        for (auto &m_op : model_outputs_->elements()){
+          outputs_.push_back(m_op.toTensor());
+        }
+  }
+  catch (std::exception& ex) {
+    try {
+      auto model_outputs_ = torch_model_->forward(inputs_).toTensor();
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Separated model run and conversion.

File: src/backends/pytorch/libtorch_backend.h [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+    std::vector<torch::jit::IValue> inputs_;
+    std::vector<torch::Tensor> outputs_;
+    torch::Device device_ = torch::Device(torch::kCPU);
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.h]: Done.

File: src/backends/pytorch/libtorch_backend.cc [hunk 26]
```diff
@@ -0,0 +1,745 @@
...
+      compute_timers.emplace_back();
+      payload.stats_->StartComputeTimer(&compute_timers.back());
+      payload.stats_->SetGPUDevice(contexts_[runner_idx]->gpu_device_);
+    }
+  }
+
+  OnCompleteQueuedPayloads(contexts_[runner_idx]->Run(this, payloads));
+
+  // clear inputs, outputs
+  contexts_[runner_idx]->inputs_.clear();
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: inputs + outputs are now local variables

File: src/backends/pytorch/libtorch_backend.cc [hunk 27]
```diff
@@ -0,0 +1,745 @@
...
+
+Status
+LibTorchBackend::Context::GetOutputTensor(
+    const std::string& name, const int& op_index, const DataType dtype, char** content,
+    size_t* byte_size, std::vector<int64_t>* content_shape)
+{
+  torch::DeviceType output_device = torch::kCPU;
+  // TODO: Fix for supporting multiple outputs
+  try {
+    outputs_[op_index] = outputs_[op_index].to(output_device);
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Will document this. In case of multiple models, the user has to ensure that the model config file contains the outputs in the same order as in the model.

File: src/backends/pytorch/libtorch_backend.h [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+
+    // Max batch size value that indicates batching is not supported.
+    static constexpr int NO_BATCHING = 0;
+
+    Context(
+        const std::string& name, const int gpu_device,
+        const int max_batch_size);
+    ~Context();
+
+    DISALLOW_COPY_AND_ASSIGN(Context);
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.h]: Also DISALLOW_MOVE(Context)

File: src/backends/pytorch/libtorch_backend.cc (hunk 17 above)
Comment (deadeyegoodwin) [reply src/backends/pytorch/libtorch_backend.cc]: There still seems to be an additional copy in GetOutputTensor...

File: src/backends/pytorch/libtorch_backend.cc [hunk 29]
```diff
@@ -0,0 +1,653 @@
...
+  torch::DeviceType output_device = torch::kCPU;
+
+  try {
+    (*outputs_)[op_index] = (*outputs_)[op_index].to(output_device);
+    torch::Tensor output_flat = (*outputs_)[op_index].flatten();
+    *byte_size = output_flat.nbytes();
+    *content = new char[*byte_size];
+
+    // Copy output into buffer
+    std::memcpy(*content, output_flat.data_ptr(), output_flat.nbytes());
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: At the very least you need to avoid this memcpy... It doesn't need to be done for this commit but file a ticket (separate from the one above) detailing the issue and what needs to be done as a performace optimization.

File: src/backends/pytorch/libtorch_backend.cc [hunk 30]
```diff
@@ -0,0 +1,653 @@
...
+Status
+LibTorchBackend::Context::GetOutputTensor(
+    std::vector<torch::Tensor>* outputs_, const int& op_index,
+    const DataType dtype, char** content, size_t* byte_size,
+    std::vector<int64_t>* content_shape)
+{
+  torch::DeviceType output_device = torch::kCPU;
+
+  try {
+    (*outputs_)[op_index] = (*outputs_)[op_index].to(output_device);
```
Comment (deadeyegoodwin) [inline src/backends/pytorch/libtorch_backend.cc]: There is no reason to first copy from GPU->CPU and then to copy from CPU->output buffer (in ReadFixedSizedOutputTensor). If output tensor is on GPU then you can just CudaMemCopy straight from there to the output buffer. Don't need to do that for this commit but file a ticket detailing what needs to be done as a performance optimization.

Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed extra copy to CPU

File: src/backends/pytorch/libtorch_backend.cc [hunk 31]
```diff
@@ -0,0 +1,653 @@
...
+
+  contexts_.emplace_back(new Context(instance_name, gpu_device, mbs));
+  Context* context = contexts_.back().get();
+
+  if (gpu_device == Context::NO_GPU_DEVICE) {
+    context->device_ = torch::Device(torch::kCPU);
+  } else {
+    context->device_ = torch::Device(torch::kCUDA, gpu_device);
+  }
+  try {
```
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Added blank line

File: src/backends/pytorch/libtorch_backend.h (hunk 28 above)
[the remark this answers is not in the document]
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.h]: Added

File: src/backends/pytorch/libtorch_backend.cc (hunk 17 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: Removed extra memcpy

File: src/backends/pytorch/libtorch_backend.cc (hunk 29 above)
Comment (CoderHam) [reply src/backends/pytorch/libtorch_backend.cc]: I fixed it here itself and avoided the extra memcpy.
